kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/RecordEncryptorTest.java

```diff
@@ -0,0 +1,298 @@
+                .singleHeader()
+                .hasKeyEqualTo("bob")
+                .hasNullValue();
+    }
+
+    @Test
+    void shouldEncryptValueOnlyPreservesNullValue() {
+        // Given
+        Set<RecordField> fields = Set.of(RecordField.RECORD_VALUE);
+        long offset = 55L;
+        long timestamp = System.currentTimeMillis();
+        var key = "hello";
+        var value = (String) null;
+
+        // When
+        var t = encryptSingleRecord(fields, offset, timestamp, key, value);
+
+        // Then
+        KafkaAssertions.assertThat(t)
+                .hasOffsetEqualTo(offset)
+                .hasTimestampEqualTo(timestamp)
+                .hasKeyEqualTo(key)
+                .hasNullValue()
+                .hasEmptyHeaders();
+
+        // And when
+        var rd = new RecordDecryptor(index -> null); // note the null return
+        var rt = transformRecord(rd, t);
+
+        // Then
+        KafkaAssertions.assertThat(rt)
+                .hasOffsetEqualTo(offset)
+                .hasTimestampEqualTo(timestamp)
+                .hasKeyEqualTo(key)
+                .hasNullValue()
+                .hasEmptyHeaders();
+    }
+
+    @Test
+    void shouldEncryptValueAndHeaders() {
+        // Given
+        Set<RecordField> fields = Set.of(RecordField.RECORD_VALUE, RecordField.RECORD_HEADER_VALUES);
+        long offset = 55L;
+        long timestamp = System.currentTimeMillis();
+        var key = "hello";
+        var value = "world";
+        var header = new RecordHeader("bob", null);
+
+        // When
+        var t = encryptSingleRecord(fields, offset, timestamp, key, value, header);
+
+        // Then
+        KafkaAssertions.assertThat(t)
+                .hasOffsetEqualTo(offset)
+                .hasTimestampEqualTo(timestamp)
+                .hasKeyEqualTo(key)
+                .hasValueNotEqualTo(value)
+                .singleHeader()
+                .hasKeyEqualTo("kroxylicious.io/encryption")
+                .hasValueEqualTo(new byte[]{ 1 });
+
+        // And when
+        var rd = new RecordDecryptor(index -> new DecryptState(t, EncryptionVersion.V1, DECRYPTOR));
+        var rt = transformRecord(rd, t);
+
+        // Then
+        KafkaAssertions.assertThat(rt)
+                .hasOffsetEqualTo(offset)
+                .hasTimestampEqualTo(timestamp)
+                .hasKeyEqualTo(key)
+                .hasValueEqualTo(value)
+                .singleHeader()
+                .hasKeyEqualTo("bob")
+                .hasNullValue();
+    }
+
+    @Test
+    @Disabled("Not implemented yet")
+    void shouldEncryptValueAndHeadersPreservesNullValue() {
+        // Given
+        Set<RecordField> fields = Set.of(RecordField.RECORD_VALUE, RecordField.RECORD_HEADER_VALUES);
+        long offset = 55L;
+        long timestamp = System.currentTimeMillis();
+        var key = "hello";
+        var value = (String) null;
+        var header = new RecordHeader("bob", null);
```

Comment:
is this an existing gap in our implementation? I was aware of the headers-only one, but can we not handle a null header value?

Reply:
I don't recall, but I guess it might be since I disabled it.

Reply:
seems to work okay in `InBandKeyManager` though there's no test

Reply:
Note: It's not the null header value that's the problem here, it's the null record value. Because it's null and we need to preserve nulls for tombstoning there's "nowhere" to put the encrypted wrapper. This is exactly the same problem as the case where we're only encrypting the value (not the headers). This test exists just so we're covering more combinations.